File: modules/02-sum-types/ocaml/lib/expr.ml
(* OCaml has coproducts as its bread and butter: a variant type IS a sum, and
   [match] is a total eliminator. A non-exhaustive match is warning 8, promoted
   to an error here (and in most CI). See the reject-ocaml project. *)

(* region:variant:start *)

type expr =
  | Lit of int
  | Add of expr * expr
  | Mul of expr * expr
  | Neg of expr

let rec eval = function
  | Lit v -> v
  | Add (l, r) -> eval l + eval r
  | Mul (l, r) -> eval l * eval r
  | Neg x -> -eval x

(* region:variant:end *)

(* [option] is itself a sum type ([None | Some of 'a]), so OCaml needs no
   nullable-pointer hack: the absence case is a distinct constructor the match
   must handle. *)
let default d = function None -> d | Some x -> x

File: modules/02-sum-types/swift/Sources/Calc/Expr.swift
// Swift enums are coproducts and `switch` must be exhaustive; a missing case is
// a compile error. `indirect` allows the recursive payload. See reject-swift.

// region:enum:start

public indirect enum Expr {
  case lit(Int)
  case add(Expr, Expr)
  case mul(Expr, Expr)
  case neg(Expr)
}

public func eval(_ e: Expr) -> Int {
  switch e {
  case .lit(let v): return v
  case .add(let l, let r): return eval(l) + eval(r)
  case .mul(let l, let r): return eval(l) * eval(r)
  case .neg(let x): return -eval(x)
  }
}

// region:enum:end

// Optional<T> is itself an enum (`.none | .some(T)`); Swift needs no nullable
// hack, and pointer-like payloads are niche-packed to the pointer's own size.
